tools/render-test/render-test-main.cpp

```diff
@@ -671,23 +655,51 @@ Result ShaderObjectRenderTestApp::writeBindingOutput(const char* fileName)
 
     for(auto outputItem : m_outputPlan.items)
     {
-        auto& inputEntry = m_shaderInputLayout.entries[outputItem.inputLayoutEntryIndex];
-        assert(inputEntry.isOutput);
-
         auto resource = outputItem.resource;
         if (resource && resource->getType() == IResource::Type::Buffer)
         {
             IBufferResource* bufferResource = static_cast<IBufferResource*>(resource.get());
-            const size_t bufferSize = bufferResource->getDesc()->sizeInBytes;
+            auto bufferDesc = *bufferResource->getDesc();
+            const size_t bufferSize = bufferDesc.sizeInBytes;
 
             ComPtr<ISlangBlob> blob;
-            m_device->readBufferResource(bufferResource, 0, bufferSize, blob.writeRef());
+            if(bufferDesc.cpuAccessFlags & IResource::AccessFlag::Read)
+            {
+                // The buffer is already allocated for CPU access, so we can read it back directly.
+                //
+                m_device->readBufferResource(bufferResource, 0, bufferSize, blob.writeRef());
+            }
+            else
+            {
+                // The buffer is not CPU-readable, so we will copy it using a staging buffer.
```

Comment:
I believe `IDevice::readBufferResource()` implementations already handle the case where the buffer is not gpu accessible and will allocate a temporary staging buffer under the hood. We can decide whether that stays in `gfx`.